packages/aws-cdk/lib/api/provisioner.ts

```diff
@@ -0,0 +1,44 @@
+import cxapi = require('@aws-cdk/cx-api');
+import { debug } from '../logging';
+import { deserializeStructure } from '../serialize';
+import { Mode } from './aws-auth/credentials';
+import { SDK } from './util/sdk';
+
+export type Template = { [key: string]: any };
+
+/**
+ * Interface for provisioners
+ *
+ * Provisioners apply templates to the cloud infrastructure.
+ */
+export interface IProvisioner {
+  readCurrentTemplate(stack: cxapi.SynthesizedStack): Promise<Template>;
```

Comment:
If this is a general purpose provisioner, that this should be something like `readCurrentState: Promise<any>` because I would also like us to view uploading to S3 and pushing to ECR as "provisioning" (or deployment tasks).

Reply:
Also related to #395

Reply:
At some point we should have generalized templates, but somewhere in the stack we should also specialize for clarity (otherwise we might as well pass `any`s around everywhere).

It's not entirely clear to me *how* this would generalize either. For example, does `readCurrentState()` for an asset deployer return the payload (wasteful) or the metadata (inconsistent with the template).

So I'd rather avoid premature generalization for now.

Reply:
For example, this could be the layer *under* the layer that deals with artifacts, in which case `readCurrentTemplate()` is perfectly appropriate.

Reply:
ok, let's leave for now, and we'll tackle soon

Reply:
For clarity: I'm only introducing the interface so that I can provide a stub for testability. The `IProvisioner` or `IDeploymentTarget` or whatever can go in a future refactor far as I'm concerned.

The germane things about the refactoring are separating out the CloudFormationProvider and adding dependency injection for testing.